File: Data_Structures/Stacks/reverse_string.py
class Stack:
    def __init__(self):
        self.stack = []
    
    def is_empty(self):
        return len(self.stack) == 0
    
    def is_full(self):
        return False
    
    def push(self, element):
        self.stack.append(element)
    
    def pop(self):
        if self.is_empty():
            print("Stack is empty. Cannot pop element.")
            return None
        return self.stack.pop()

def reverse_string(string):
    stack = Stack()

    for char in string:
        stack.push(char)

    for i in range(len(string)):
        string[i] = stack.pop()

    return string

string = list("Hello, World!")
print("Original string:", "".join(string))

reversed_string = reverse_string(string)
print("Reversed string:", "".join(reversed_string))
print("\n")

File: Data_Structures/Stacks/reverse_string.c
// Write a program to reverse a given string using a stack. The input string can contain any characters

#include <stdio.h>
#include <string.h>
#define MAX_SIZE 100

typedef struct {
    char stack[MAX_SIZE];
    int top;
} Stack;

void initialize(Stack* stack) {
    stack->top = -1;
}

int isEmpty(Stack* stack) {
    return stack->top == -1;
}

int isFull(Stack* stack) {
    return stack->top == MAX_SIZE - 1;
}

void push(Stack* stack, char element) {
    if (isFull(stack)) {
        printf("Stack is full. Cannot push element.\n");
        return;
    }
    stack->stack[++stack->top] = element;
}

char pop(Stack* stack) {
    if (isEmpty(stack)) {
        printf("Stack is empty. Cannot pop element.\n");
        return '\0';
    }
    return stack->stack[stack->top--];
}

void reverseString(char* string) {
    int len = strlen(string);
    Stack stack;
    initialize(&stack);

    for (int i = 0; i < len; i++)
        push(&stack, string[i]);

    for (int i = 0; i < len; i++)
        string[i] = pop(&stack);
}

int main() {
    char string[] = "Hello, World!";
    printf("Original string: %s\n", string);

    reverseString(string);
    printf("Reversed string: %s\n", string);
    printf("\n");

    return 0;
}
